pkg/bgpv1/test/adverts_test.go

```diff
@@ -0,0 +1,360 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	slim_metav1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+	slimv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+	"github.com/cilium/cilium/pkg/testutils"
+
+	"github.com/stretchr/testify/require"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	// maxTestDuration is allowed time for test execution
+	maxTestDuration = 15 * time.Second
+)
+
+// Test_PodCIDRAdvert validates pod IPv4 subnet is advertised, withdrawn and modified on node addresses change.
+func Test_PodCIDRAdvert(t *testing.T) {
+	testutils.PrivilegedTest(t)
+
+	// steps define order in which test is run. Note, this is different from table tests, in which each unit is
+	// independent. In this case, tests are run sequentially and there is dependency on previous test step.
+	var steps = []struct {
+		description         string
+		podCIDRs            []string
+		expectedRouteEvents []routeEvent
+	}{
+		{
+			description: "advertise pod CIDRs",
+			podCIDRs: []string{
+				"10.1.0.0/16",
+				"aaaa::/64",
+			},
+			expectedRouteEvents: []routeEvent{
+				{
+					sourceASN:   ciliumASN,
+					prefix:      "10.1.0.0",
+					prefixLen:   16,
+					isWithdrawn: false,
+				},
+				{
+					sourceASN:   ciliumASN,
+					prefix:      "aaaa::",
+					prefixLen:   64,
+					isWithdrawn: false,
+				},
+			},
+		},
+		{
+			description: "delete pod CIDRs",
+			podCIDRs:    []string{},
+			expectedRouteEvents: []routeEvent{
+				{
+					sourceASN:   ciliumASN,
+					prefix:      "10.1.0.0",
+					prefixLen:   16,
+					isWithdrawn: true,
+				},
+				{
+					sourceASN:   ciliumASN,
+					prefix:      "aaaa::",
+					prefixLen:   64,
+					isWithdrawn: true,
+				},
+			},
+		},
+		{
+			description: "re-add pod CIDRs",
+			podCIDRs: []string{
+				"10.1.0.0/16",
+				"aaaa::/64",
+			},
+			expectedRouteEvents: []routeEvent{
+				{
+					sourceASN:   ciliumASN,
+					prefix:      "10.1.0.0",
+					prefixLen:   16,
+					isWithdrawn: false,
+				},
+				{
+					sourceASN:   ciliumASN,
+					prefix:      "aaaa::",
+					prefixLen:   64,
+					isWithdrawn: false,
+				},
+			},
+		},
+		{
+			description: "update pod CIDRs",
+			podCIDRs: []string{
+				"10.2.0.0/16",
+				"bbbb::/64",
+			},
+			expectedRouteEvents: []routeEvent{
+				{
+					sourceASN:   ciliumASN,
+					prefix:      "10.1.0.0",
+					prefixLen:   16,
+					isWithdrawn: true,
+				},
+				{
+					sourceASN:   ciliumASN,
+					prefix:      "10.2.0.0",
+					prefixLen:   16,
+					isWithdrawn: false,
+				},
+				{
+					sourceASN:   ciliumASN,
+					prefix:      "aaaa::",
+					prefixLen:   64,
+					isWithdrawn: true,
+				},
+				{
+					sourceASN:   ciliumASN,
+					prefix:      "bbbb::",
+					prefixLen:   64,
+					isWithdrawn: false,
+				},
+			},
+		},
+	}
```

Comment:
How can we choose the test case covered in this component test? For example, we can test this re-advertise case with the reconciler's unit test. How did you choose this case?

Reply:
This component level test is mainly BDD ( behavior driven development ), where high level features are tested. 

So, what I see following tests to be added 
- Pod CIDR advertisements
- LB Service advertisements
- Neighbor behavior
    - Authentication
    - Timers
    - Graceful restart

Some special test cases
- Restarts 
- Any customer reported bugs and issues with order of operations. 

I think overall, we need to add unit tests in reconciler, controller and gobgp shim layer while doing the development. Whereas, under test/ directory, we should have high level tests which show case BGP features.

Reply:
I think reconciler test is doing bit beyond it's role, it should be basically testing just diffs which is the logical processing reconciler is doing. I think validating gobgp internals from reconciler is leaking underlying abstraction of gobgp into generic reconciler code.

Reply:
Strictly speaking, BDD would require some sort of formal language for defining test cases - like gherkin (godog). 

Using go tests provide more flexibility, but could be difficult to read if steps are not defined clearly. I will add this in the description.